src/sets.jl

```diff
@@ -438,6 +438,34 @@ Base.isapprox(a::SOS2{T}, b::SOS2{T}; kwargs...) where T = isapprox(a.weights, b
 
 dimension(s::Union{SOS1, SOS2}) = length(s.weights)
 
+"""
+	ActivationCond
+
+Activation condition for an indicator constraint.
+Used as first type parameter of `IndicatorSet{ActiveCond,S}`.
+"""
+@enum ActivationCond begin
+    ACTIVATE_ON_FALSE
+    ACTIVATE_ON_TRUE
+end
+
+"""
+    IndicatorSet{ActiveCond, S <: AbstractScalarSet}
+
+Set of (x,y) that satisfy the indicator constraint:
+    y ∈ 𝔹, y = ActiveCond ==> a^T x ∈ S
+
+`S` has to be an `AbstractScalarSet`.
+`ActiveCond` is one of the value of the `ActivationCond` enum.
+`IndicatorSet` is used with a `VectorAffineFunction` holding
+the indicator variable first.
+"""
+struct IndicatorSet{ActiveCond, S <: AbstractScalarSet} <: AbstractVectorSet
+    s::S
+    IndicatorSet{ActiveCond}(s::S) where {ActiveCond, S <: AbstractScalarSet} = new{ActiveCond,S}(s)
+end
```

Comment:
I haven't tested, but you probably want something like the following
 - use meaningful names where possible, so `s` becomes `set`
 - `ActiveCond` should be restricted to the right type
 - use escaped latex in the docstring so that it renders properly

```julia
"""
	ActivationCondition

Activation condition for an indicator constraint.
Used as first type parameter of [`IndicatorSet`](@ref).
"""
@enum ActivationCondition begin
    ACTIVATE_ON_FALSE
    ACTIVATE_ON_TRUE
end

"""
    IndicatorSet{A, S <: AbstractScalarSet}

The set ``\\{(y, x) \\in \\{0, 1\\} \\times \\mathbb{R}: y = A \\implies x \\in set\\}``,
where `A` is an [`ActivationCondition`](@ref) enum.

`IndicatorSet` is used with an [`AbstractVectorFunction`](@ref) of dimension 2,
with the first element being the indicator variable `y`, and the second element
being a scalar function of `x`.
"""
struct IndicatorSet{A, S <: AbstractScalarSet} <: AbstractVectorSet
    set::S
    function IndicatorSet(A::ActivationCondition, set::AbstractScalarSet)
        return new{A, typeof(set)}(set)
    end
end
```

There is a debate about whether `set` should be an `AbstractScalarSet`. I don't see why it shouldn't be an `AbstractVectorSet` with one less dimension than the function.

Reply:
Is there any solver that natively supports the equivalent of indicator constraints for VAF? Why do we need this extra complexity when you could represent the same thing by adding multiple indicator constraints with a common indicator variable?

Reply:
> y = A ... where `A` is an [`ActivationCondition`](@ref) enum.
Doesn't make much sense. `y` doesn't take enum values. I'd explicitly list the cases, e.g.,
```
The set ``\\{(y, x) \\in \\{0, 1\\} \\times \\mathbb{R}: y = 0 \\implies x \\in set\\}``
when `A` is `ACTIVATE_WHEN_ZERO` and
``\\{(y, x) \\in \\{0, 1\\} \\times \\mathbb{R}: y = 1 \\implies x \\in set\\}``
when `A` is `ACTIVATE_WHEN_ONE`.

Reply:
> Is there any solver that natively supports the equivalent of indicator constraints for VAF? Why do we need this extra complexity when you could represent the same thing by adding multiple indicator constraints with a common indicator variable?

Adding multiple indicator variables does not work if the set is `SecondOrderCone` and modeling with such set may be useful even if it is not supported by any solver if it could be bridge. However I am not sure how to bridge a vector set.
For a scalar set you can add a variable `u` and add the constraints `u + x in set` and `[y, u]  in SOS1` but it does not work if `u` is a vector as you might need several components to be nonzero.

Reply:
@odow 
> ActiveCond should be restricted to the right type

(replaced by `A` in the new version), it's not a type but an enum value, I don't think you can check type parameter type (just like `Array{T,N}` does not check that `N` is an integer

Reply:
also, I'm waiting for a final design agreement to add tests

Reply:
> Adding multiple indicator variables does not work if the set is SecondOrderCone and modeling with such set may be useful even if it is not supported by any solver if it could be bridge. 

Good point. But I propose we address the problem at hand and keep `S` as a scalar set. If, in the future, we want indicator constraints on vector sets, we can add them. We already have redundant ways to express scalar sets using 1-dimensional vector sets, so having a scalar indicator constraint and a vector indicator constraint would be consistent with MOI.